core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala

```diff
@@ -0,0 +1,260 @@
+
+import scala.jdk.CollectionConverters._
+
+object BrokerMetadataListener{
+  val MetadataBatchProcessingTimeUs = "MetadataBatchProcessingTimeUs"
+  val MetadataBatchSizes = "MetadataBatchSizes"
+}
+
+class BrokerMetadataListener(val brokerId: Int,
+                             val time: Time,
+                             val metadataCache: RaftMetadataCache,
+                             val configRepository: LocalConfigRepository,
+                             val groupCoordinator: GroupCoordinator,
+                             val replicaManager: RaftReplicaManager,
+                             val txnCoordinator: TransactionCoordinator,
+                             val logManager: LogManager,
+                             val threadNamePrefix: Option[String]
+                            ) extends MetaLogListener with KafkaMetricsGroup {
+  val logContext = new LogContext(s"[BrokerMetadataListener id=${brokerId}] ")
+  val log = logContext.logger(classOf[BrokerMetadataListener])
+  logIdent = logContext.logPrefix()
+
+  /**
+   * A histogram tracking the time in microseconds it took to process batches of events.
+   */
+  private val batchProcessingTimeHist = newHistogram(BrokerMetadataListener.MetadataBatchProcessingTimeUs)
+
+  /**
+   * A histogram tracking the sizes of batches that we have processed.
+   */
+  private val metadataBatchSizeHist = newHistogram(BrokerMetadataListener.MetadataBatchSizes)
+
+  /**
+   * The highest metadata offset that we've seen.  Written only from the event queue thread.
+   */
+  @volatile private var _highestMetadataOffset = -1L
+
+  val eventQueue = new KafkaEventQueue(time, logContext, threadNamePrefix.getOrElse(""))
+
+  def highestMetadataOffset(): Long = _highestMetadataOffset
+
+  /**
+   * Handle new metadata records.
+   */
+  override def handleCommits(lastOffset: Long, records: util.List[ApiMessage]): Unit = {
+    eventQueue.append(new HandleCommitsEvent(lastOffset, records))
+  }
+
+  class HandleCommitsEvent(lastOffset: Long,
+                           records: util.List[ApiMessage])
+      extends EventQueue.FailureLoggingEvent(log) {
+    override def run(): Unit = {
+      if (isDebugEnabled) {
+        debug(s"Metadata batch ${lastOffset}: handling ${records.size()} record(s).")
+      }
+      val imageBuilder =
+        MetadataImageBuilder(brokerId, log, metadataCache.currentImage())
+      val startNs = time.nanoseconds()
+      var index = 0
+      metadataBatchSizeHist.update(records.size())
+      records.iterator().asScala.foreach { case record =>
+        try {
+          if (isTraceEnabled) {
+            trace("Metadata batch %d: processing [%d/%d]: %s.".format(lastOffset, index + 1,
+              records.size(), record.toString()))
+          }
+          handleMessage(imageBuilder, record, lastOffset)
+        } catch {
+          case e: Exception => error(s"Unable to handle record ${index} in batch " +
+            s"ending at offset ${lastOffset}", e)
+        }
```

Comment:
Perhaps not to resolve now, but I think error conditions like this might need more thought.  Should the broker die?  If so, how would the broker ever come back up if it can't get past a particular batch?  But what is the impact of not processing the remaining messages in a batch, and is that always acceptable or just sometimes acceptable?  Do we need metrics to track these error events?  It seems an error log message might not be sufficient?